### PR TITLE
feat:Support duplicate one existing Route

### PR DESCRIPTION
**Pull request by batman-ezio**
Please answer these questions before submitting a pull request

Why submit this pull request?
 - [ ] Bugfix
 - [x] New feature provided
 - [ ] Improve performance
 - [ ]  Backport patches
 - [ ] We has added two fields for upstream in APISIX.

Support duplicate one existing Route  https://github.com/apache/apisix-dashboard/issues/1487

New feature or improvement
1: duplicate button from route list
2: check duplicate name from POST route api

Describe the details and related test reports.

1:Click duplicate button from the router list, the address will be  `routes/:rid/duplicate`
image

2: update the name to make sure it's no duplicated
image

3: submit data, return to the list. The new route will be there

image

### CODE REVIEW

File: api/internal/handler/route/route.go
Comment (nic-chen) [inline api/internal/handler/route/route.go:311]: should add this check for `Update` too. Thanks.

**Comment by nic-chen** [comment]
need some test cases for duplicate check, thanks. @batman-ezio

File: api/internal/handler/route/route.go
Comment (batman-ezio) [reply api/internal/handler/route/route.go:311]: ok

**Comment by batman-ezio** [comment]
> need some test cases for duplicate check, thanks. @batman-ezio

@nic-chen after I check duplicate name in create/update route api. some test cases failed. I think they are use the same route name to test before.  some errors throw from the route test are very strange , like this 
```
=== RUN   TestRoute_Create/create_route_success
panic: 
assert: mock: I don't know what to return because the method call was unexpected.
	Either do Mock.On("List").Return(...) first, or remove the List() call.
	This method was unexpected:
		List(store.ListInput)
		0: store.ListInput{Predicate:(func(interface {}) bool)(0x11bf5a0), Format:(func(interface {}) interface {})(nil), PageSize:0, PageNumber:0, Less:(func(interface {}, interface {}) bool)(nil)}
	at: [store_mock.go:35 route.go:537 route.go:311 route_test.go:1005] [recovered]
	panic: 
assert: mock: I don't know what to return because the method call was unexpected.
	Either do Mock.On("List").Return(...) first, or remove the List() call.
	This method was unexpected:
		List(store.ListInput)
		0: store.ListInput{Predicate:(func(interface {}) bool)(0x11bf5a0), Format:(func(interface {}) interface {})(nil), PageSize:0, PageNumber:0, Less:(func(interface {}, interface {}) bool)(nil)}
	at: [store_mock.go:35 route.go:537 route.go:311 route_test.go:1005]

``` 
I have change the route name but still got this. can you help?

**Comment by nic-chen** [comment]
> > need some test cases for duplicate check, thanks. @batman-ezio
> 
> @nic-chen after I check duplicate name in create/update route api. some test cases failed. I think they are use the same route name to test before. some errors throw from the route test are very strange , like this
> 
> ```
> === RUN   TestRoute_Create/create_route_success
> panic: 
> assert: mock: I don't know what to return because the method call was unexpected.
> 	Either do Mock.On("List").Return(...) first, or remove the List() call.
> 	This method was unexpected:
> 		List(store.ListInput)
> 		0: store.ListInput{Predicate:(func(interface {}) bool)(0x11bf5a0), Format:(func(interface {}) interface {})(nil), PageSize:0, PageNumber:0, Less:(func(interface {}, interface {}) bool)(nil)}
> 	at: [store_mock.go:35 route.go:537 route.go:311 route_test.go:1005] [recovered]
> 	panic: 
> assert: mock: I don't know what to return because the method call was unexpected.
> 	Either do Mock.On("List").Return(...) first, or remove the List() call.
> 	This method was unexpected:
> 		List(store.ListInput)
> 		0: store.ListInput{Predicate:(func(interface {}) bool)(0x11bf5a0), Format:(func(interface {}) interface {})(nil), PageSize:0, PageNumber:0, Less:(func(interface {}, interface {}) bool)(nil)}
> 	at: [store_mock.go:35 route.go:537 route.go:311 route_test.go:1005]
> ```
> 
> I have change the route name but still got this. can you help?


hi @batman-ezio , you need to mock `routeStore.List` for  `checkDuplicateName`

you could refer to:
https://github.com/apache/apisix-dashboard/blob/2de2c72cfa8899e98045c3059d281b90cf544404/api/internal/handler/plugin_config/plugin_config_test.go#L694-L709

**Comment by Jaycean** [comment]
@batman-ezio 
Hi, I see that in your E2E test and unit test, when creating a route, the error code returned is 400, which is usually the wrong data format or type of the route. Please check the data format or type.

**Comment by batman-ezio** [comment]
> @batman-ezio
> Hi, I see that in your E2E test and unit test, when creating a route, the error code returned is 400, which is usually the wrong data format or type of the route. Please check the data format or type.

hi @Jaycean  I found some test case use the same route name or empty route name for e2e test,  test cases begin to bump up after my fix one by one

**Comment by Jaycean** [comment]
> > @batman-ezio
> > Hi, I see that in your E2E test and unit test, when creating a route, the error code returned is 400, which is usually the wrong data format or type of the route. Please check the data format or type.
> 
> hi @Jaycean I found some test case use the same route name or empty route name for e2e test, test cases begin to bump up after my fix one by one
Yes, 400 is usually caused by wrong data format or type. In E2E test, a route ID r1 is usually created and passed, and the route with ID r1 will be deleted, so there is no problem with the same name.

**Comment by nic-chen** [comment]
@batman-ezio  would you mind if we modify your code on your branch?

**Comment by batman-ezio** [comment]
> @batman-ezio would you mind if we modify your code on your branch?

no, please
I found those test case  in `ece/route_with_*`  does not pass Name field in the body, so their route will create with a blank  name and will failed for duplicate, can we add a logic to create/update to set the name from url params if  name is not in body or blank?

**Comment by nic-chen** [comment]
> can we add a logic in create/update API to set the name from url params if it does not exists in the request body?

oh, no, we should not do that.
we could fix in test cases which do duplicate. I think there will not be a lot, because most of test cases will delete the routes at the end.

**Comment by batman-ezio** [comment]
> > can we add a logic in create/update API to set the name from url params if it does not exists in the request body?
> 
> oh, no, we should not do that.
> we could fix in test cases which do duplicate. I think there will not be a lot, because most of test cases will delete the routes at the end.

ok, can you help to fix the rest test cases, thanks

**Comment by nic-chen** [comment]
> > > can we add a logic in create/update API to set the name from url params if it does not exists in the request body?
> > 
> > 
> > oh, no, we should not do that.
> > we could fix in test cases which do duplicate. I think there will not be a lot, because most of test cases will delete the routes at the end.
> 
> ok, can you help to fix the rest test cases, thanks


sure, my pleasure. will update later.

**Comment by batman-ezio** [comment]
> > > > can we add a logic in create/update API to set the name from url params if it does not exists in the request body?
> > > 
> > > 
> > > oh, no, we should not do that.
> > > we could fix in test cases which do duplicate. I think there will not be a lot, because most of test cases will delete the routes at the end.
> > 
> > 
> > ok, can you help to fix the rest test cases, thanks
> 
> sure, my pleasure. will update later.

thanks. I have squashed all my test case commits into one commit, and rebased from master (which includes the latest frontend e2e fix), but the test case still failed

**Comment by nic-chen** [comment]
> > > > > can we add a logic in create/update API to set the name from url params if it does not exists in the request body?
> > > > 
> > > > 
> > > > oh, no, we should not do that.
> > > > we could fix in test cases which do duplicate. I think there will not be a lot, because most of test cases will delete the routes at the end.
> > > 
> > > 
> > > ok, can you help to fix the rest test cases, thanks
> > 
> > 
> > sure, my pleasure. will update later.
> 
> thanks. I have squashed all my test case commits into one commit, and rebased from master (which includes the latest frontend e2e fix), but the test case still failed

I will try to fix backend CI this afternoon.
@LiteSun  could you please help for the FE?

**Comment by LiteSun** [comment]
> I will try to fix backend CI this afternoon.
> @LiteSun could you please help for the FE?

ok

**Comment by nic-chen** [comment]
backend CI had been resolved.
please help check FE CI when you have time. @LiteSun

**Comment by membphis** [comment]
ping @LiteSun

**Comment by batman-ezio** [comment]
after rebase from latest master, the frontend test case is still failed. 

`should import route(s) from be test files`
failed because from click multi routes.  maybe import in foreach look needs to ensure the route is deleted. add async?

**Comment by batman-ezio** [comment]
> ping @LiteSun

I see he fire the PR for fix, https://github.com/apache/apisix-dashboard/pull/1590/files

**Review by starsz — APPROVED** [review]
BackEnd LGTM. The same things should be done in other module.

**Comment by membphis** [comment]
@batman-ezio you can make a try to merge the `master` branch

**Comment by LiteSun** [comment]
> @batman-ezio you can make a try to merge the `master` branch

Just help you sync master branch code. let's wait for CI report 😊

**Comment by starsz** [comment]
Wait for CI.

**Comment by juzhiyuan** [comment]
Why backend failed?

**Comment by nic-chen** [comment]
> Why backend failed?

need some time to check for it

**Comment by LiteSun** [comment]
ping

**Comment by nic-chen** [comment]
@batman-ezio  please sync codes from branch master

**Comment by batman-ezio** [comment]
@nic-chen  This PR contains two things
1: duplicate a route 
2: check duplicate name in create/update route api

#2 break many test case and take much time. 
how about I revert the code and just keep the `#1` codes here . 

I see you have a more effective  PR to do the `#2` things  here  https://github.com/apache/apisix-dashboard/pull/1606/files

and then we can move this PR forward

**Comment by nic-chen** [comment]
> @nic-chen This PR contains two things
> 1: duplicate a route
> 2: check duplicate name in create/update route api
> 
> `#2` break many test case and take much time.
> how about I revert the code and just keep the `#1` codes here .
> 
> I see you have a more effective PR to do the `#2` things here https://github.com/apache/apisix-dashboard/pull/1606/files
> 
> and then we can move this PR forward

OK

**Comment by batman-ezio** [comment]
CI is still not stable, I sync from master still failes

**Comment by nic-chen** [comment]
hi @batman-ezio  
pr #1606  has been merged. please sync codes from branch master.

File: web/config/routes.ts
Comment (juzhiyuan) [inline web/config/routes.ts:43]: Do we need this path?

File: web/src/pages/Route/List.tsx
Comment (juzhiyuan) [inline web/src/pages/Route/List.tsx:390]: Could we fetch data first, then submit it directly by using `create` method?

Comment (batman-ezio) [reply web/src/pages/Route/List.tsx:390]: we can't do that because the route name is unique, we need to take user to `/routes/:rid/duplicate` and fill the copied data, user needs to change the name and then get next to submit

Comment (juzhiyuan) [reply web/src/pages/Route/List.tsx:390]: ok, got it

File: web/config/routes.ts
Comment (batman-ezio) [reply web/config/routes.ts:43]: > hi @batman-ezio
> pr #1606 has been merged. please sync codes from branch master.

sync from master but frontend still fails

Comment (juzhiyuan) [reply web/config/routes.ts:43]: cc @guoqqqi to help resolve failed test

Comment (guoqqqi) [reply web/config/routes.ts:43]: Hi, I have fixed the problem with the relevant test and wait for #1608.

Comment (batman-ezio) [reply web/config/routes.ts:43]: @guoqqqi  I sync from master after your PR merged, but still failed

File: web/cypress/integration/route/create-edit-duplicate-delete-route.spec.js
Comment (LiteSun) [inline web/cypress/integration/route/create-edit-duplicate-delete-route.spec.js:173]: To make the test case stable, it would be better to close the notification here.

File: web/cypress/integration/route/create-edit-duplicate-delete-route.spec.js
Comment (guoqqqi) [inline web/cypress/integration/route/create-edit-duplicate-delete-route.spec.js:169]:
```suggestion
      cy.get(domSelector.name).clear().type(routeName);
```
This will fix the CI error.

**Comment by nic-chen** [comment]
please resolve conflicts

**Comment by batman-ezio** [comment]
> please resolve conflicts

fixed

**Review by LiteSun — APPROVED** [review]

File: web/cypress/integration/route/create-edit-duplicate-delete-route.spec.js
Comment (juzhiyuan) [inline web/cypress/integration/route/create-edit-duplicate-delete-route.spec.js:188]: @LiteSun Do we need to visit this page by clicking Sidebar Menu?

**Review by liuxiran — APPROVED** [review]
LGTM， thanks

and there is a UE  enhancement suggestion: 
when copying the route user in the page of first step and click [Next] button directly, they will get `duplicate route name`error, maybe leaving the route name blank would be a better choice.

**Review by juzhiyuan — APPROVED** [review]

**Comment by juzhiyuan** [comment]
ok, then let me merge.

**Comment by juzhiyuan** [comment]
thanks! @batman-ezio

**Pull request merged**
